### ISSUE TITLE
[FEATURE] LetsEncrypt certificates for Harvester

### ISSUE THREAD

**Issue by Roguito** [harvester/harvester#3425]
[QUESTION|FEATURE] I am wondering what is the best way to apply LetsEncrypt certificates for harvester. I can apply them manually via the UI but as they expire after 90 days I'd rather find an automagic way to perform this.

Any guidance on kube cronjob or integrated cert-manager solution... can this be done on the base appliance, hidden in the private k3s/rancher/CNI cluster, can be communicated to via API, etc?

**Comment by zatricky** [comment]
This seems like a fundamental security/compliance issue :-|

**Comment by mortenn** [comment]
a cert-manager addon would be a nice addition

**Comment by ER-EPR** [comment]
> a cert-manager addon would be a nice addition

Agree! Is there a work round to implement a certificate update script, kind of thing now, before the Add-on is developed? 
Why nobody cares about this, I think it should be a basic function.

**Comment by claywd** [comment]
Yeah I could be interested in building this. Would prefer if certs on the load balancer and load balancer itself was configured via crd in k8s so I can build a harvester-issuer for cert-manager that follows existing patterns vs trying to update settings on the harvester nodes individually. 

Questions:
Are people thinking this is just for the harvester instance itself? 
Did people want individual certs added for load balancers in harvester? 
Does harvester currently terminate tls for all load balancer listeners? If so, does it assume that the domain will always be the same as the harvester cluster and thus use the same cert? 

Didn't even see that the settings were configurable via api. Without access to the crd I'd be executing provledged scripts to update the ssl cert. Not really palatable.

**Comment by johnwc** [comment]
The load-balancer in harvester is not a layer-7 load balancer, it is a layer-4 load balancer. They forward the raw packets to the nodes, so currently there cannot be certificates at the load balancer. I have brought this issue up prior, as there is also no liveness checks able to be configured other than checking if the node is listening on the port. Another existing issue with the way the current load balancer is running, is that it directs all traffic to one node. It causes a bottle neck as well as causes small issues when it needs to transfer the listening load balancer to another node, as it uses arp to advertise where the packets should be directed. You can peruse through the LB setup [here](https://github.com/harvester/load-balancer-harvester). It uses kube-vip to do the load balancing. I do think there is room for improvement.

It would be nice if the load-balancer supported layer-7 so that we could do tls termination with acme ssl certificates.

**Comment by claywd** [comment]
Oh geeze. Well layer 4 means we can configure our certs on another layer 7 load balancer behind the harvester layer 4 load balancer and terminate tls there at least. Idk, maybe I prefer it that way rather than reconstructing the harvester load balancer to also support layer 7.

So what terminates tls for the harvester ui? Is that reconfigurable? Ie if I wanted to handle certs for it using cert manager and envoy then can I just update harvester to direct traffic to the layer 7 behind the layer 4? Seems like doing so and configuring cert manager in Co junction with the layer 7 load balancer as well as adding an entry for the harvester dashboard behind it would solve the problem op is describing.

Any other endpoints would just be cert request and lb7 entries.

**Comment by claywd** [comment]
Thinking contour is a good candidate. https://projectcontour.io/getting-started/

Going to need to figure out updating the harvester/rancher ui endpoint so it points to the l7 load balancer.

**Comment by johnwc** [comment]
At the end of the day, the configuration of kube-vip for the load balancer needs to be updated or extended so that we can configure it manually with CRDs or similar. All traffic going to only one node is not ideal and causes a bottleneck. It's not a load balancer in its current configuration, but rather a failover. 

I believe the harvester UI is terminated on the node that is currently hosting the harvester VIP. You could easily host a nginx reverse proxy that forwards requests to the harvester VIP, then setup an acme certificate on the nginx. The point your DNS for your harvester FQDN to the nginx instance. Similar to how you'd setup a rancher HA cluster.

**Comment by claywd** [comment]
@johnwc I agree with that! Crds to configure the layer 4 lb is a must. Curious though, why not use metallb? It already does this right? Just extend it and put it behind the Ui no?

**Comment by silug** [comment]
> I believe the harvester UI is terminated on the node that is currently hosting the harvester VIP. You could easily host a nginx reverse proxy that forwards requests to the harvester VIP, then setup an acme certificate on the nginx. The point your DNS for your harvester FQDN to the nginx instance. Similar to how you'd setup a rancher HA cluster.

I did this for a while, but at the time the `Open in WebVNC` links would not work (or, rather, they would open a new window that would never load content).

**Comment by johnwc** [comment]
> I did this for a while, but at the time the `Open in WebVNC` links would not work (or, rather, they would open a new window that would never load content).

If the webvnc used gRPC or websockets, then nginx would have to be setup to accommodate for those as well. We had a similar issue where a product was using standard http/rest calls for most of the site, but gRPC for small part.

**Comment by silug** [comment]
> If the webvnc used gRPC or websockets, then nginx would have to be setup to accommodate for those as well. We had a similar issue where a product was using standard http/rest calls for most of the site, but gRPC for small part.

You are 100% correct.  I just fixed the nginx configuration, and now it works. :slightly_smiling_face: 

https://nginx.org/en/docs/http/websocket.html
